frontend/packages/dev-console/src/components/pipelines/pipeline-builder/update-utils.ts

```diff
@@ -128,11 +128,11 @@ const getErrors = (task: PipelineTask, resource: PipelineResourceTask): TaskErro
   const resources = getTaskResources(resource);
 
   const taskInputResources = task.resources?.inputs?.length || 0;
-  const requiredInputResources = resources.inputs?.length || 0;
+  const requiredInputResources = (resources.inputs || []).filter((r) => !r?.optional).length;
   const missingInputResources = requiredInputResources - taskInputResources > 0;
 
   const taskOutputResources = task.resources?.outputs?.length || 0;
-  const requiredOutputResources = resources.outputs?.length || 0;
+  const requiredOutputResources = (resources.outputs || []).filter((r) => !r?.optional).length;
```

Comment:
Not sure why you didn't just add `?.filter(...)` instead of defaulting to an empty array. Not that it really matters.